### PR TITLE
JB-36: Implement domain model of Payment for JobBot server

### PR DESCRIPTION
**Pull request by fitzroy-c**
Todo:
Implement domain model entity for JobBot server

- [x] Job

- [x] Outcome

- [x] Payment

- [x] Account

with unit tests

### CODE REVIEW

**Comment by SomezakiNayuki** [comment]
Review not completed, still reviewing. But good job and thanks for the effort. Please raise any question with me I'll explain it.

File: JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Job.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    @NotNull
+    private Date time; // post time?
+
+    private Date expiredDate;
+
+    private JobStatusEnum jobStatusEnum;
+
+    @OneToOne
+    @JoinColumn(name = "user_id")
```
Comment (SomezakiNayuki) [inline JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Job.java:39]: `name = "employer_id"`.

File: JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Job.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    private User employer;
+
+    @OneToOne // to be clarified
+    @JoinColumn(name = "user_id")
```
Comment (SomezakiNayuki) [inline JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Job.java:43]: `name = "employee_id"`.

File: JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Payment.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+public class Payment {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private long transactionId;
+
+    private double amount;
+
+    @OneToOne
+    @JoinColumn(name = "user_id")
```
Comment (SomezakiNayuki) [inline JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Payment.java:20]: `name = "payer_id"`.

File: JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Payment.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    private User payer;
+
+    @OneToOne
+    @JoinColumn(name = "user_id")
```
Comment (SomezakiNayuki) [inline JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Payment.java:24]: `name = "payee_id"`.

File: JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/ProductImage.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+@Entity
+public class ProductImage {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    private String imageUrl;
+
+    @ManyToOne
+    @JoinColumn(name = "outcome_images")
```
Comment (SomezakiNayuki) [inline JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/ProductImage.java:14]: `name = "outcome_id"`.

File: JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/ProductImage.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package org.dizzybot.jobbot.entities;
+
+import jakarta.persistence.*;
+
+@Entity
+public class ProductImage {
```
Comment (SomezakiNayuki) [inline JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/ProductImage.java:6]: getter & setter?

File: JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/User.java [hunk 7]
```diff
@@ -30,24 +35,32 @@ public class User {
...
     public UserProfile userProfile;
 
-//    public List<Job> jobPosted = new ArrayList<Job>();
-//
-//    public List<Job> jobAccepted = new ArrayList<Job>();
-//
-//    @Nullable
-//    public Account account;
+
+    @OneToMany(mappedBy = "user", cascade = CascadeType.ALL)
```
Comment (SomezakiNayuki) [inline JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/User.java:39]: `mappedBy = "employer"`.

File: JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/User.java [hunk 8]
```diff
@@ -30,24 +35,32 @@ public class User {
...
+    public List<Job> jobPosted = new ArrayList<Job>();
+
+    @OneToMany(mappedBy = "user", cascade = CascadeType.ALL)
```
Comment (SomezakiNayuki) [inline JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/User.java:42]: `mappedBy = "employee"`.

File: JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/enums/PaymentStatusEnum.java [hunk 9]
```diff
@@ -0,0 +1,8 @@
+package org.dizzybot.jobbot.enums;
+
+public enum PaymentStatusEnum {
+    UNPAID,
+    SUCCESSFULLY_PAID,
+    ABNORMAL;
+
```
Comment (SomezakiNayuki) [inline JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/enums/PaymentStatusEnum.java:7]: Minor: remove empty line.

File: JobBot-server/jobbot/jobbot/src/test/java/org/dizzybot/jobbot/entities/AccountTest.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    private User user;
+
+    private Account account;
+
+    private Validator validator = Validation.buildDefaultValidatorFactory().getValidator();
+
+    @Before
+    public void beforeEach() {
+        this.user = new User("Dizzy Bot", "Password", "dizzybot@email.com");
+        this.account = this.user.account;
```
Comment (SomezakiNayuki) [inline JobBot-server/jobbot/jobbot/src/test/java/org/dizzybot/jobbot/entities/AccountTest.java:26]: `this.user.getAccount()`, do not directly access attribute, access attribute using getter.

File: JobBot-server/jobbot/jobbot/src/test/java/org/dizzybot/jobbot/entities/AccountTest.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+    @After
+    public void afterEach() {
+        this.user = null;
+        this.account = null;
+    }
+
+    @Test
+    public void testUserAccountCreateByUser() {
+        assertNotNull(this.user.account);
```
Comment (SomezakiNayuki) [inline JobBot-server/jobbot/jobbot/src/test/java/org/dizzybot/jobbot/entities/AccountTest.java:37]: `this.user.getAccount()`, do not directly access attribute, access attribute using getter.

File: JobBot-server/jobbot/jobbot/src/test/java/org/dizzybot/jobbot/entities/JobTest.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    public void afterEach() {
+        this.jobAccepted = null;
+        this.jobPosted = null;
+    }
+
+    @Test
+    public void testUserJobListsCreateByUser() {
+        assertNotNull(this.user.jobAccepted);
+        assertNotNull(this.user.jobPosted);
+    }
```
Comment (SomezakiNayuki) [inline JobBot-server/jobbot/jobbot/src/test/java/org/dizzybot/jobbot/entities/JobTest.java:43]: Do not directly access attribute, access attribute using getter.

File: JobBot-server/jobbot/jobbot/src/test/java/org/dizzybot/jobbot/entities/PaymentTest.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    private User userEmployee;
+
+    @Before
+    public void beforeEach() {
+        this.userEmployee = new User("Dizzy Bot Employee", "Password", "dizzybotemployee@email.com");
+        this.userEmployer = new User("Dizzy Bot Employer", "Password", "dizzybotemployer@email.com");
+        this.userEmployer.setAccount(new Account(15468779464L));
+        this.payment = new Payment(123.0, this.userEmployer, this.userEmployee);
+        this.payment_0 = new Payment(321, this.userEmployer, this.userEmployee);
+    }
```
Comment (SomezakiNayuki) [inline JobBot-server/jobbot/jobbot/src/test/java/org/dizzybot/jobbot/entities/PaymentTest.java:27]: Do not directly access attribute, access attribute using getter.

File: JobBot-server/jobbot/jobbot/src/test/java/org/dizzybot/jobbot/entities/JobTest.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    private Job jobTest;
+
+    private Validator validator = Validation.buildDefaultValidatorFactory().getValidator();
+
+    @Before
+    public void beforeEach() {
+        this.user = new User("Dizzy Bot", "Password", "dizzybot@email.com");
+        this.jobAccepted = this.user.jobAccepted;
+        this.jobPosted = this.user.jobPosted;
+    }
```
Comment (SomezakiNayuki) [inline JobBot-server/jobbot/jobbot/src/test/java/org/dizzybot/jobbot/entities/JobTest.java:31]: Do not directly access attribute, access attribute using getter.

**Review by SomezakiNayuki — CHANGES_REQUESTED** [review]
Modify based on comments.

File: JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Account.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+@Setter
+public class Account {
+
+    @NonNull
+    private long cardNumber; // should have some constrains
+
+    @Nullable
+    private long bsb;
+
+    @Id
```
[the remark this answers is not in the document]
Comment (fitzroy-c) [reply JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Account.java:25]: currently, add this unique constraint annotation to the class level and set a new attribute id as Id (long). hope will fix this

File: JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Payment.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    private User payee;
+
+    private PaymentStatusEnum status;
+
+    @ManyToOne
+    @JoinColumn(name = "account_paymentHistory")
```
[the remark this answers is not in the document]
Comment (fitzroy-c) [reply JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Payment.java:30]: so the one after @joincolumn should be the table u wanna connect and the related foreign key in that table, which in this case account_id. In this situation, will the attribute in Account object automatically add the new payment with its index in the paymentHistory attribute?

File: JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Job.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+    private User employee;
+
+    @OneToOne
+    @JoinColumn(name = "outcome_id")
+    private Outcome outcome;
+
+    @ManyToOne
+    @JoinColumn(name = "user_jobPosted")
+    private User userPost; // todo
```
[the remark this answers is not in the document]
Comment (fitzroy-c) [reply JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Job.java:52]: my former concern is that @many2one and @one2many should appear in pairs. that is the reason I added these 2

File: JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Job.java (hunk 1 above)
Comment (fitzroy-c) [reply JobBot-server/jobbot/jobbot/src/main/java/org/dizzybot/jobbot/entities/Job.java:39]: so it is the name of the instance, not the class ~.~

File: JobBot-server/jobbot/jobbot/src/test/java/org/dizzybot/jobbot/entities/JobTest.java (hunk 12 above)
Comment (fitzroy-c) [reply JobBot-server/jobbot/jobbot/src/test/java/org/dizzybot/jobbot/entities/JobTest.java:43]: this is to test the set/get function or is it better to access attributes this way~

File: JobBot-server/jobbot/jobbot/src/test/java/org/dizzybot/jobbot/entities/PaymentTest.java (hunk 13 above)
Comment (fitzroy-c) [reply JobBot-server/jobbot/jobbot/src/test/java/org/dizzybot/jobbot/entities/PaymentTest.java:27]: currently changed to 
    @Before
    public void beforeEach() {
        this.payee = new User(...);
        this.payer = new User(...);
        this.payer.setAccount(new Account(15468779464L));
        this.payment = new Payment(123.0, this.payer, this.payee);
        this.payment_0 = new Payment(321, this.payer, this.payee);
    } this is already the User instance that should be put in, so I don't know how to use getter here...

**Review by SomezakiNayuki — APPROVED** [review]
Good job

**Review by SomezakiNayuki — APPROVED** [review]

**Pull request merged**